README.md
nv-jdo
======

Overview
--------

Utilities for JDO.

| Class          | Description                                     |
|:---------------|:------------------------------------------------|
| `Dao`          | Generic DAO implementation based on JDO.        |
| `TaskAdapter`  | Empty implementation of `TransactionAwareTask`. |
| `TaskExecutor` | Task executor with or without transaction.      |

| Interface              | Description                                                      |
|:-----------------------|:-----------------------------------------------------------------|
| `Task`                 | Task executed by `TaskExecutor`.                                 |
| `TransactionAwareTask` | Sub interface of `Task` with additional methods for transaction. |


License
-------

Apache License, Version 2.0


Maven
-----

```xml
<dependency>
    <groupId>com.neovisionaries</groupId>
    <artifactId>nv-jdo</artifactId>
    <version>1.19</version>
</dependency>
```


Source Code
-----------

  <code>https://github.com/TakahikoKawasaki/nv-jdo.git</code>


Javadoc
-------

  <code>http://TakahikoKawasaki.github.io/nv-jdo/</code>


Example
-------

```java
//------------------------------------------------------
// Example of Dao, which is a generic DAO implementation
// based on JDO.
//------------------------------------------------------

// Persistence manager factory.
PersistenceManagerFactory factory = ...;

// Create a DAO with an entity class and the persistence manager factory.
Dao<Customer> dao = new Dao<Customer>(Customer.class, factory);

// Or using a static method.
Dao<Customer> dao = Dao.create(Customer.class, factory);

// Use the DAO.

// (1) Create an entity.
Customer customer = ...;
dao.put(customer);

// (2) Get an entity.
String customerId = ...;
Customer customer = dao.getById(customerId);

// (3) Delete an entity.
dao.delete(customer);

// (4) Delete an entity by ID.
dao.deleteById(customerId);
```

```java
//------------------------------------------------------
// Example of TaskExecutor.
//------------------------------------------------------
// Persistence manager factory.
PersistenceManagerFactory factory = ...;

// Create a task executor.
TaskExecutor executor = new TaskExecutor(factory);

// Create a task.
Task task = new TaskAdapter() {
    @Override
    public void beforeTransactionBegin(PersistenceManager manager, Transaction tx) {
        tx.setOptimistic(false);
    }

    @Override
    public Object run(PersistenceManager manager) {
        ......
    }
};

// Execute the task.
executor.execute(task, true, 2);
```


See Also
--------

* [JDO API 3.0](http://db.apache.org/jdo/api30/apidocs/index.html)


Author
------

[Authlete, Inc.](https://www.authlete.com/)<br/>
Takahiko Kawasaki &lt;taka@authlete.com&gt;


Contributors
------------

[Authlete, Inc.](https://www.authlete.com/)<br/>
Hideki Ikeda &lt;hide@authlete.com&gt;


CHANGES.md
CHANGES
=======

1.19 (2022-05-30)
-----------------

- `TaskExecutor` class
    * Added `TaskExecutor(PersistenceManager)` constructor.
    * Added `execute(Task, int)` method.
    * Added `getFactory()` method.
    * Added `getManager()` method.


1.18 (2018-07-09)
-----------------

- Added `executeListWithArrayByQuery` methods to `Dao` class.
- Added `executeListWithMapByQuery` methods to `Dao` class.


1.17 (2017-08-03)
-----------------

- Added `QueryLanguage` enum.
- Added `executeWithArrayByQuery` methods to `Dao` class.
- Added `executeWithMapByQuery` methods to `Dao` class.


1.16 (2017-04-15)
-----------------

- Added `TransactionAwareTask` interface.
- Added `TaskAdapter` class.
- Updated `TaskExecutor` to handle `TransactionAwareTask`.


src/main/java/com/neovisionaries/jdo/Task.java
/*
 * Copyright (C) 2014-2017 Neo Visionaries Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.neovisionaries.jdo;


import javax.jdo.PersistenceManager;


/**
 * Task executed by {@link TaskExecutor}.
 *
 * <p>
 * {@link TransactionAwareTask} is an interface which inherits this interface.
 * {@link TaskAdapter} is an empty implementation of the {@code TransactionAwareTask}
 * interface.
 * </p>
 *
 * @since 1.2
 *
 * @author Takahiko Kawasaki
 */
public interface Task
{
    /**
     * Task body.
     *
     * @param manager
     *         A persistence manager.
     *
     * @return
     *         Any arbitrary object. The object returned by
     *         this method is returned from {@link TaskExecutor}'s
     *         {@code execute} methods. {@code TaskExecutor} does
     *         not care about what the object is. Even {@code null}
     *         is OK.
     */
    public Object run(PersistenceManager manager);
}


src/main/java/com/neovisionaries/jdo/JdoDao.java
/*
 * Copyright (C) 2014 Neo Visionaries Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.neovisionaries.jdo;


import javax.jdo.JDOHelper;
import javax.jdo.JDOObjectNotFoundException;
import javax.jdo.JDOUserException;
import javax.jdo.PersistenceManager;
import javax.jdo.PersistenceManagerFactory;


/**
 * DAO based on JDO.
 *
 * <style type="text/css">
 * pre.sample span.comment { color: #009900; }
 * </style>
 * <pre class="sample" style="border: solid 1px black; margin: 0.5em; padding: 0.5em;">
 * <span class="comment">// Set up a shared persistence manager factory.</span>
 * JdoDao.{@link #setupSharedPersistenceManagerFactory(String)
 * setupSharedPersistenceManagerFactory}("transactions-optional");
 *
 * <span class="comment">// Create a DAO with an entity class.</span>
 * JdoDao&lt;Customer&gt; dao = new JdoDao&lt;Customer&gt;(Customer.class);
 *
 * <span class="comment">// or</span>
 * JdoDao&lt;Customer&gt; dao = JdoDao.{@link #create(Class) create}(Customer.class);
 *
 * <span class="comment">// Use the DAO.</span>
 * Customer customer = dao.{@link #getById(Object) getById}(customerId);
 * </pre>
 *
 * @param <TEntity>
 */
public class JdoDao<TEntity>
{
    private static PersistenceManagerFactory sharedPersistenceManagerFactory;
    private final Class<TEntity> entityClass;
    private PersistenceManagerFactory persistenceManagerFactory;


    /**
     * Get the persistence manager factory that has been set by
     * {@link #setSharedPersistenceManagerFactory(PersistenceManagerFactory)}.
     *
     * @return
     *         The shared persistence manager factory.
     */
    public static PersistenceManagerFactory getSharedPersistenceManagerFactory()
    {
        return sharedPersistenceManagerFactory;
    }


    /**
     * Set a persistence manager factory that is shared as fail-safe for
     * {@code JdoDao} instances which do not have a persistence manager
     * factory of their own.
     *
     * @param factory
     *         A persistence manager factory.
     */
    public static void setSharedPersistenceManagerFactory(PersistenceManagerFactory factory)
    {
        sharedPersistenceManagerFactory = factory;
    }


    /**
     * Set up a shared persistence manager factor. This method calls
     * {@link JDOHelper#getPersistenceManagerFactory(String)
     * JDOHelper.getPersistenceManagerFactory}{@code (name)} and sets
     * the returned value as the shared persistence manager factory.
     *
     * <pre style="border: solid 1px black; margin: 0.5em; padding: 0.5em;">
     * JdoDao.{@link #setupSharedPersistenceManagerFactory(String)
     * setupSharedPersistenceManagerFactory}("transactions-optional");</pre>
     *
     * @param name
     */
    public static void setupSharedPersistenceManagerFactory(String name)
    {
        sharedPersistenceManagerFactory = JDOHelper.getPersistenceManagerFactory(name);
    }


    /**
     * Create a DAO for the specified entity.
     *
     * <p>
     * This method just does the following:
     * </p>
     *
     * <blockquote><pre>
     * return new JdoDao&lt;TEntity&gt;(entityClass);
     * </pre></blockquote>
     *
     * @param entityClass
     *
     * @return
     *         A new DAO.
     *
     * @since 1.1
     */
    public static <TEntity> JdoDao<TEntity> create(Class<TEntity> entityClass)
    {
        return new JdoDao<TEntity>(entityClass);
    }


    /**
     * Constructor.
     *
     * @param entityClass
     *         The class of entities handled by this DAO.
     */
    public JdoDao(Class<TEntity> entityClass)
    {
        this(entityClass, null);
    }


    /**
     * Constructor with a persistence manager factory for this instance.
     *
     * @param entityClass
     *         The class of entities handled by this DAO.
     *
     * @param factory
     *         A persistence manager factory.
     */
    public JdoDao(Class<TEntity> entityClass, PersistenceManagerFactory factory)
    {
        checkNonNull(entityClass, "entityClass");

        this.entityClass = entityClass;
        this.persistenceManagerFactory = factory;
    }


    /**
     * Get the persistence manager factory for this instance.
     *
     * @return
     *         The persistence manager factory that has been set by
     *         {@link #setPersistenceManagerFactory(PersistenceManagerFactory)}.
     */
    public PersistenceManagerFactory getPersistenceManagerFactory()
    {
         return persistenceManagerFactory;
    }


    /**
     * Set a persistence manager factory for this instance.
     *
     * @param factory
     *         A persistence manager factory to set.
     *
     * @return
     *         {@code this} object.
     */
    public JdoDao<TEntity> setPersistenceManagerFactory(PersistenceManagerFactory factory)
    {
        this.persistenceManagerFactory = factory;

        return this;
    }


    /**
     * Get a persistence manager, which is obtained from the persistence
     * manager factory which is set to this JdoDao instance, or from the
     * shared persistence manager factory when no persistence manager
     * factory is set to this instance.
     *
     * @return
     *         A persistence manager, or {@code null} if no persistence
     *         manager factory is set.
     */
    public PersistenceManager getPersistenceManager()
    {
        PersistenceManagerFactory factory = getPersistenceManagerFactory();

        if (factory == null)
        {
            factory = sharedPersistenceManagerFactory;
        }

        if (factory == null)
        {
            return null;
        }

        return factory.getPersistenceManager();
    }


    private PersistenceManager getPersistenceManagerAndCheck()
    {
        PersistenceManager manager = getPersistenceManager();

        if (manager == null)
        {
            throw new IllegalStateException("Persistence Manager is not available.");
        }

        return manager;
    }


    private void checkNonNull(Object object, String name)
    {
        if (object == null)
        {
            throw new IllegalArgumentException(name + " is null.");
        }
    }


    /**
     * Delete an entity. This method calls
     * {@link PersistenceManager#deletePersistent(Object)}.
     *
     * @param entity
     *         The entity to delete.
     *
     * @throws IllegalArgumentException
     *         {@code entity} is {@code null}.
     *
     * @throws IllegalStateException
     *         Persistent Manager is not available.
     */
    public void delete(TEntity entity)
    {
        checkNonNull(entity, "entity");

        PersistenceManager manager = getPersistenceManagerAndCheck();

        try
        {
            manager.deletePersistent(entity);
        }
        finally
        {
            manager.close();
        }
    }


    /**
     * Delete an entity. This method calls
     * {@link PersistenceManager#getObjectById(Class, Object)}
     * to find the entity and then calls
     * {@link PersistenceManager#deletePersistent(Object)}
     * to delete the entity.
     *
     * @param id
     *         The ID of the entity.
     *
     * @throws IllegalArgumentException
     *         {@code id} is {@code null}.
     *
     * @throws IllegalStateException
     *         Persistent Manager is not available.
     */
    public void deleteById(Object id)
    {
        checkNonNull(id, "id");

        PersistenceManager manager = getPersistenceManagerAndCheck();

        try
        {
            TEntity entity = manager.getObjectById(entityClass, id);

            if (entity != null)
            {
                manager.deletePersistent(entity);
            }
        }
        finally
        {
            manager.close();
        }
    }


    /**
     * Get an entity. This method calls
     * {@link PersistenceManager#getObjectById(Class, Object)}.
     *
     * @param id
     *         The ID of the entity.
     *
     * @return
     *         The entity, or {@code null} if not found.
     *
     * @throws IllegalArgumentException
     *         {@code id} is {@code null}.
     *
     * @throws IllegalStateException
     *         Persistent Manager is not available.
     */
    public TEntity getById(Object id)
    {
        checkNonNull(id, "id");

        PersistenceManager manager = getPersistenceManagerAndCheck();

        try
        {
            return manager.getObjectById(entityClass, id);
        }
        catch (JDOObjectNotFoundException e)
        {
            // Not found.
            return null;
        }
        finally
        {
            manager.close();
        }
    }


    /**
     * Make an entity persistent. This method calls
     * {@link PersistenceManager#makePersistent(Object)}.
     *
     * @param entity
     *         An entity to be persistent.
     *
     * @throws IllegalArgumentException
     *         {@code entity} is {@code null}.
     *
     * @throws IllegalStateException
     *         Persistent Manager is not available.
     *
     * @throws JDOUserException
     */
    public void put(TEntity entity)
    {
        checkNonNull(entity, "entity");

        PersistenceManager manager = getPersistenceManagerAndCheck();

        try
        {
            manager.makePersistent(entity);
        }
        finally
        {
            manager.close();
        }
    }
}


src/main/java/com/neovisionaries/jdo/TaskAdapter.java
/*
 * Copyright (C) 2017 Neo Visionaries Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.neovisionaries.jdo;


import javax.jdo.PersistenceManager;
import javax.jdo.Transaction;


/**
 * Empty implementation of the {@link TransactionAwareTask} interface.
 *
 * @since 1.16
 *
 * @author Takahiko Kawasaki
 */
public class TaskAdapter implements TransactionAwareTask
{
    public Object run(PersistenceManager manager)
    {
        return null;
    }


    public void beforeTransactionBegin(PersistenceManager manager, Transaction tx)
    {
    }


    public void afterTransactionBegin(PersistenceManager manager, Transaction tx)
    {
    }


    public void beforeTransactionCommit(PersistenceManager manager, Transaction tx)
    {
    }


    public void afterTransactionCommit(PersistenceManager manager, Transaction tx)
    {
    }


    public void beforeTransactionRollback(PersistenceManager manager, Transaction tx)
    {
    }


    public void afterTransactionRollback(PersistenceManager manager, Transaction tx)
    {
    }
}


src/main/java/com/neovisionaries/jdo/QueryLanguage.java
/*
 * Copyright (C) 2017 Neo Visionaries Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific
 * language governing permissions and limitations under the
 * License.
 */
package com.neovisionaries.jdo;


/**
 * Query language.
 *
 * @since 1.17
 *
 * @author Takahiko Kawasaki
 */
public enum QueryLanguage
{
    /**
     * JDOQL; The identifier is "{@code javax.jdo.query.JDOQL}".
     */
    JDOQL("javax.jdo.query.JDOQL"),


    /**
     * SQL; The identifier is "{@code javax.jdo.query.SQL}".
     */
    SQL("javax.jdo.query.SQL"),
    ;


    /**
     * The identifier of this query language.
     */
    private final String identifier;


    /**
     * Private constructor with the identifier of this query language.
     *
     * @param identifier
     *         The identifier of this query language.
     */
    private QueryLanguage(String identifier)
    {
        this.identifier = identifier;
    }


    /**
     * Get the identifier of this query language. For example,
     * {@link #JDOQL}.{@code getIdentifier()} returns {@code
     * "javax.jdo.query.JDOQL"}.
     *
     * @return
     *         The identifier of this query language.
     */
    public String getIdentifier()
    {
        return identifier;
    }
}


src/main/java/com/neovisionaries/jdo/TaskExecutor.java
/*
 * Copyright (C) 2014-2022 Neo Visionaries Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.neovisionaries.jdo;


import javax.jdo.JDOCanRetryException;
import javax.jdo.JDOException;
import javax.jdo.PersistenceManager;
import javax.jdo.PersistenceManagerFactory;
import javax.jdo.Transaction;


/**
 * Task executor.
 *
 * <style type="text/css">
 * pre.sample span.comment { color: darkgreen; }
 * </style>
 * <pre class="sample" style="border: solid 1px black; margin: 0.5em; padding: 0.5em;">
 * <span class="comment">// Persistence manager factory.</span>
 * {@link PersistenceManagerFactory} factory = ...;
 *
 * <span class="comment">// Create a task executor.</span>
 * {@link TaskExecutor} executor = new {@link #TaskExecutor(PersistenceManagerFactory) TaskExecutor}{@code (factory)};
 *
 * <span class="comment">// Create a task.</span>
 * {@link Task} task = new {@link TaskAdapter}() {
 *     &#x40;Override
 *     public void beforeTransactionBegin({@link PersistenceManager} manager, {@link Transaction} tx) {
 *         tx.{@link Transaction#setOptimistic(boolean) setOptimistic}(false);
 *     }
 *
 *     &#x40;Override
 *     public Object run({@link PersistenceManager} manager) {
 *         ......
 *     }
 * };
 *
 * <span class="comment">// Execute the task.</span>
 * executor.{@link #execute(Task, boolean, int) execute}{@code (task, true, 2)};
 * </pre>
 *
 * @since 1.2
 *
 * @author Takahiko Kawasaki
 */
public class TaskExecutor
{
    private PersistenceManagerFactory mFactory;
    private PersistenceManager mManager;


    /**
     * Constructor with a persistence manager factory.
     *
     * @param factory
     *         A persistence manager factory. Must not be {@code null}.
     *
     * @throws IllegalArgumentException
     *         {@code factory} is {@code null}.
     */
    public TaskExecutor(PersistenceManagerFactory factory)
    {
        // Check if the factory is null.
        if (factory == null)
        {
            // Factory must not be null.
            throw new IllegalArgumentException("factory is null.");
        }

        this.mFactory = factory;
    }


    /**
     * Constructor with a persistence manager.
     *
     * @param manager
     *         A persistence manager. Must not be {@code null}.
     *
     * @throws IllegalArgumentException
     *         {@code manager} is {@code null}.
     *
     * @since 1.19
     */
    public TaskExecutor(PersistenceManager manager)
    {
        // Check if the manager is null.
        if (manager == null)
        {
            // Manager must not be null.
            throw new IllegalArgumentException("manager is null.");
        }

        this.mManager = manager;
    }


    /**
     * Execute a task.
     *
     * <p>
     * This method is an alias of {@link #execute(Task, boolean, int)
     * execute(task, false, 0)}.
     * </p>
     *
     * @param task
     *         Task to be executed.
     *
     * @return
     *         The object returned from {@code task.}{@link Task#run(PersistenceManager)
     *         run(PersistenceManager)}.
     *
     * @throws IllegalArgumentException
     *         {@code task} is {@code null}.
     *
     * @throws JDOException
     *         A persistence manager could not be obtained or the task failed.
     *         See the description of {@link #execute(Task, boolean, int)}
     *         for details.
     */
    public Object execute(Task task)
    {
        return execute(task, false, 0);
    }


    /**
     * Execute a task.
     *
     * <p>
     * This method is an alias of {@link #execute(Task, boolean, int)
     * execute(task, transaction, 0)}.
     * </p>
     *
     * @param task
     *         Task to be executed.
     *
     * @param transaction
     *         {@code true} to execute the task in a transaction.
     *
     * @return
     *         The object returned from {@code task.}{@link Task#run(PersistenceManager)
     *         run(PersistenceManager)}.
     *
     * @throws IllegalArgumentException
     *         {@code task} is {@code null}.
     *
     * @throws JDOException
     *         A persistence manager could not be obtained or the task failed.
     *         See the description of {@link #execute(Task, boolean, int)}
     *         for details.
     */
    public Object execute(Task task, boolean transaction)
    {
        return execute(task, transaction, 0);
    }


    /**
     * Execute a task.
     *
     * <p>
     * This method is an alias of {@link #execute(Task, boolean, int)
     * execute(task, false, retryCount)}.
     * </p>
     *
     * @param task
     *         Task to be executed.
     *
     * @param retryCount
     *         Indicates how many times to retry the task. The task is retried
     *         only when {@code task.}{@link Task#run(PersistenceManager)
     *         run(PersistenceManager)} threw a {@link JDOCanRetryException}.
     *         If an exception of other type was thrown, the task is not retried.
     *
     * @return
     *         The object returned from {@code task.}{@link Task#run(PersistenceManager)
     *         run(PersistenceManager)}.
     *
     * @throws IllegalArgumentException
     *         {@code task} is {@code null}, or {@code retryCount} is less than 0.
     *
     * @throws JDOException
     *         A persistence manager could not be obtained or the task failed.
     *         See the description of {@link #execute(Task, boolean, int)}
     *         for details.
     *
     * @since 1.19
     */
    public Object execute(Task task, int retryCount)
    {
        return execute(task, false, retryCount);
    }


    /**
     * Execute a task.
     *
     * <p>
     * If this {@code TaskExecutor} instance was created by the constructor
     * {@link #TaskExecutor(PersistenceManagerFactory)}, a new instance of
     * {@link PersistenceManager} is created inside this {@code execute} method
     * by calling the {@link PersistenceManagerFactory#getPersistenceManager()
     * getPersistenceManager()} method of the {@link PersistenceManagerFactory}
     * instance that was passed via the constructor.
     * </p>
     *
     * <p>
     * On the other hand, if this {@code TaskExecutor} instance was created by
     * the constructor {@link #TaskExecutor(PersistenceManager)}, the
     * {@link PersistanceManager} instance passed via the constructor is used
     * inside this {@code execute} method.
     * </p>
     *
     * <p>
     * Note that if the current transaction of the {@link PersistenceManager}
     * instance that was passed via the constructor
     * {@link #TaskExecutor(PersistenceManager)} is already active, transaction
     * operations are NOT executed even if the {@code transaction} argument is
     * {@code true}.
     * </p>
     *
     * @param task
     *         Task to be executed. If the given instance implements the
     *         {@link TransactionAwareTask} interface, transaction-related
     *         callback methods of the interface will be called appropriately.
     *
     * @param transaction
     *         {@code true} to execute the task in a transaction.
     *
     * @param retryCount
     *         Indicates how many times to retry the task. The task is retried
     *         only when {@code task.}{@link Task#run(PersistenceManager)
     *         run(PersistenceManager)} threw a {@link JDOCanRetryException}.
     *         If an exception of other type was thrown, the task is not retried.
     *
     * @return
     *         The object returned from {@code task.}{@link Task#run(PersistenceManager)
     *         run(PersistenceManager)}.
     *
     * @throws IllegalArgumentException
     *         {@code task} is {@code null}, or {@code retryCount} is less than 0.
     *
     * @throws JDOException
     *         There are several cases for this exception.
     *         <ul>
     *         <li>Failed to get a persistence manager from the persistence manager factory.
     *         <li>{@link PersistenceManagerFactory#getPersistenceManager()
     *             getPersistenceManager()} of the persistence manager factory
     *             returned {@code null}.
     *         <li>The task continued to throw {@link JDOCanRetryException} and
     *             exceeded the retry count.
     *         </ul>
     *
     * @throws RuntimeException
     *         The task continued to throw {@link JDOCanRetryException} and
     *         exceeded the retry count.
     */
    public Object execute(Task task, boolean transaction, int retryCount)
    {
        // Check if the task is null.
        if (task == null)
        {
            // Task must not be null.
            throw new IllegalArgumentException("task is null.");
        }

        // Check if the retry count is less than 0.
        if (retryCount < 0)
        {
            // Retry count must not be less than 0.
            throw new IllegalArgumentException("retryCount < 0");
        }

        // If this TaskExecutor instance was created by TaskExecutor(PersistenceManager),
        // the PersistenceManager instance given via the constructor is used.
        //
        // Otherwise, a new PersistenceManager instance is created here by calling the
        // getPersistenceManager() method of the PersistenceManagerFactory instance
        // that was given via TaskExecutor(PersistenceManagerFactory).
        PersistenceManager manager = (mManager != null) ? mManager : createManager();
        boolean managerCreated     = (mManager == null);

        try
        {
            // Execute the task.
            return doExecute(task, transaction, retryCount, manager);
        }
        finally
        {
            if (managerCreated)
            {
                // Close the persistence manager in any case.
                manager.close();
            }
        }
    }


    /**
     * Get the persistence manager factory that was given to the
     * {@link #TaskExecutor(PersistenceManagerFactory)} constructor.
     *
     * @return
     *         The instance of {@link PersistenceManagerFactory} that was given
     *         to the {@link #TaskExecutor(PersistenceManagerFactory)} constructor.
     *         {@code null} is returned if this {@code TaskExecutor} instance was
     *         created by other constructors.
     *
     * @since 1.19
     */
    public PersistenceManagerFactory getFactory()
    {
        return mFactory;
    }


    /**
     * Get the persistence manager that was given to the
     * {@link #TaskExecutor(PersistenceManager)} constructor.
     *
     * @return
     *         The instance of {@link PersistenceManager} that was given
     *         to the {@link #TaskExecutor(PersistenceManager)} constructor.
     *         {@code null} is returned if this {@code TaskExecutor} instance was
     *         created by other constructors.
     *
     * @since 1.19
     */
    public PersistenceManager getManager()
    {
        return mManager;
    }


    private PersistenceManager createManager()
    {
        PersistenceManager manager = null;

        try
        {
            // Get a persistence manager from the persistence manager factory.
            manager = mFactory.getPersistenceManager();
        }
        catch (JDOException e)
        {
            // getPersistenceManager() threw an exception.
            throw e;
        }
        catch (RuntimeException e)
        {
            // getPersistenceManager() threw an unexpected exception.
            throw new JDOException("getPersistenceManager() failed.", e);
        }

        if (manager == null)
        {
            // Failed to obtain a persistence manager.
            throw new JDOException("getPersistenceManager() returned null.");
        }

        return manager;
    }


    private Object doExecute(Task task, boolean transaction, int retryCount, PersistenceManager manager)
    {
        Transaction tx = manager.currentTransaction();

        // A flag that indicates whether to do transaction operations.
        // If the current transaction is already active, transaction
        // operations are not executed even if the 'transaction' flag
        // passed to the 'execute' method is true.
        boolean doTx = transaction && !tx.isActive();

        // If transaction is required.
        if (doTx)
        {
            // Begin a transaction.
            beginTransaction(manager, task, tx);
        }

        try
        {
            // Execute the task
            Object ret = doExecute(task, retryCount, manager);

            if (doTx)
            {
                // Commit the changes.
                commitTransaction(manager, task, tx);
            }

            return ret;
        }
        finally
        {
            // If transaction is required but not committed.
            if (doTx && tx.isActive())
            {
                // Revoke the changes.
                rollbackTransaction(manager, task, tx);
            }
        }
    }


    private static TransactionAwareTask cast(Task task)
    {
        if (task instanceof TransactionAwareTask)
        {
            return (TransactionAwareTask)task;
        }
        else
        {
            return null;
        }
    }


    private static void beginTransaction(PersistenceManager manager, Task task, Transaction tx)
    {
        // Cast by TransactionAwareTask if possible.
        TransactionAwareTask tat = cast(task);

        if (tat != null)
        {
            // Call a hook before beginning a transaction.
            tat.beforeTransactionBegin(manager, tx);
        }

        // Start a transaction.
        // begin() throws JDOUserException if something is wrong.
        tx.begin();

        if (tat != null)
        {
            // Call a hook after beginning a transaction.
            tat.afterTransactionBegin(manager, tx);
        }
    }


    private static void commitTransaction(PersistenceManager manager, Task task, Transaction tx)
    {
        // Cast by TransactionAwareTask if possible.
        TransactionAwareTask tat = cast(task);

        if (tat != null)
        {
            // Call a hook before committing a transaction.
            tat.beforeTransactionCommit(manager, tx);
        }

        // Commit the changes.
        tx.commit();

        if (tat != null)
        {
            // Call a hook after committing a transaction.
            tat.afterTransactionCommit(manager, tx);
        }
    }


    private static void rollbackTransaction(PersistenceManager manager, Task task, Transaction tx)
    {
        // Cast by TransactionAwareTask if possible.
        TransactionAwareTask tat = cast(task);

        if (tat != null)
        {
            // Call a hook before rollbacking a transaction.
            tat.beforeTransactionRollback(manager, tx);
        }

        // Revoke the changes.
        tx.rollback();

        if (tat != null)
        {
            // Call a hook after rollbacking a transaction.
            tat.afterTransactionRollback(manager, tx);
        }
    }


    private Object doExecute(Task task, int retryCount, PersistenceManager manager)
    {
        // The exception that task.run(manager) threw.
        JDOCanRetryException lastException = null;

        // Retries up to the retry count at a maximum.
        for (int count = 0; count <= retryCount; ++count)
        {
            try
            {
                // Execute the task.
                return task.run(manager);
            }
            catch (JDOCanRetryException e)
            {
                // The task threw an exception, but it may succeed when retried.
                lastException = e;
            }
        }

        // The task continued to throw JDOCanRetryException and
        // exceeded the retry count.
        throw lastException;
    }
}
